File: wk4/java/jollyjumpers.java
// problem link: https://open.kattis.com/problems/jollyjumpers

import java.io.*;
import java.math.*;
import java.util.*;

public class jollyjumpers {
    public static void main(String[] args) throws IOException {
        BufferedReader r = new BufferedReader(new InputStreamReader(System.in));
        String s;
        while((s = r.readLine()) != null) {
            int[] a = Arrays.stream(s.split(" ")).mapToInt(Integer::parseInt).toArray();
            HashSet<Integer> set = new HashSet<>();
            boolean flag = true;
            for(int i = 2; i < a.length && flag; i++) {
                int c = Math.abs(a[i] - a[i - 1]);
                if(c == 0 || c > a[0] - 1 || set.contains(c)) {
                    System.out.println("Not jolly");
                    flag = false;
                }
                set.add(c);
            }
            if(flag)
                System.out.println("Jolly");
        }
    }
}

File: wk4/cpp/jollyjumpers.cpp
// problem link: https://open.kattis.com/problems/jollyjumpers

#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
    
	int n;
	while(cin >> n) {
	    vector<bool> l(n, false);
	    int p;
	    cin >> p;
	    bool f = true;
	    for(int i = 1; i < n; i++) {
	        int c;
	        cin >> c;
	        if(!f)
	            continue;
	        if(c == p || abs(c - p) >= n || l[abs(c - p)])
	            f = false;
	        else {
	            l[abs(c - p)] = true;
	            p = c;
	        }
	    }
	    cout << (f ? "Jolly\n" : "Not jolly\n");
	}
	    
	return 0;
}